### ISSUE TITLE
Clarify when a DAP should respond to requests like terminateRequest... immediately, or when the request has been handled?

### ISSUE THREAD

**Issue by DanTup** [microsoft/debug-adapter-protocol#367]
I've always assumed that if the DAP client issues a request like `terminate`, that the DAP server should try to perform the termination before responding to this request (since this allows it to communicate a failure, and if the response was just always sent immediately before any work is done, it seems like it's essentially an event and doesn't need request/response).

However, I've noticed that if I don't respond to a terminate request within 2s, the VS Code client shows an error message that the termination timed out (even though it didn't, and the DAP server is still working on it). This happens a lot in test runs, because the Dart test runner will by default wait to allow the current test to complete when first trying to stop (to provide more consistency for `teardowns`, not having to have them deal with tests that were aborted mid-running).

I can change how this works to respond immediately, but I can't find anything in the docs that describe that this is how servers should behave, and it seems like something that should be (since if clients and servers interpret this differently - as has happened here - users may see confusing timeout messages).

**Comment by puremourning** [comment]
Seems like the client just has an overly aggressive timeout. I agree with your interpretation- client should wait for response. 2s is (sadly) not very long in debugger timescales in 2023

**Comment by DanTup** [comment]
Yeah, I generally think timeouts (particular short ones!) are a bad idea anyway. Give the user a progress notification and a cancel option and let _them_ decide when to give up. The number of times I've known something would take a long time and I'm happy to go and make a cup of tea while it does it, but some arbitrary timeout aborts it when it's nearly done 🤷‍♂️

I'll hold off making changes to my DAP for now though in the top VS Code can be updated to not be so aggressive.

**Comment by connor4312** [comment]
I agree that the preferred option here is a tweak to VS Code behavior. The response should be sent once termination completes.
